google/cloud/spanner/interval.cc

```diff
@@ -0,0 +1,367 @@
+    std::int64_t minute;
+    if (!ParseInteger(t, minute, false)) return false;
+    d += absl::Minutes(minute);
+    if (!absl::ConsumePrefix(&t, ":")) return false;
+    double second;
+    if (ParseDouble(t, second, false)) {
+      d += absl::Seconds(second);
+    } else {
+      std::int64_t second;
+      if (!ParseInteger(t, second, false)) return false;
+      d += absl::Seconds(second);
+    }
+    intvl = Interval(sign == "-" ? -d : d);
+    s.remove_prefix(s.size() - t.size());
+    return true;
+  }
+
+  static Interval Unit(absl::string_view& s, std::int32_t n) {
+    auto t = s;
+    auto unit = ConsumeWord(t);
+    if (unit.size() > 1) absl::ConsumeSuffix(&unit, "s");  // unpluralize
+    for (auto const& factory : kUnitFactories) {
+      if (absl::StartsWith(factory.name, unit)) {
+        s.remove_prefix(s.size() - t.size());
+        return factory.factory(n);
+      }
+    }
+    return Interval(0, 0, n);  // default to days without removing unit
+  }
+
+  std::string s_;
+};
+
+StatusOr<Interval> Parser::Parse() {
+  Interval intvl;
+  auto s = absl::string_view(s_);
+
+  ConsumeSpaces(s);
+  if (absl::ConsumePrefix(&s, "@")) {
+    ConsumeSpaces(s);
+  }
+
+  for (;; ConsumeSpaces(s)) {
+    Interval vi;
+    if (ParseYearMonth(s, vi)) {
+      intvl += vi;
+      continue;
+    }
+    if (ParseHourMinuteSecond(s, vi)) {
+      intvl += vi;
+      continue;
+    }
+    double vf;
+    if (ParseDouble(s, vf, true)) {
+      ConsumeSpaces(s);
+      intvl += Unit(s, 1) * vf;
+      continue;
+    }
+    std::int32_t vn;
+    if (ParseInteger(s, vn, true)) {
+      ConsumeSpaces(s);
+      intvl += Unit(s, vn);
+      continue;
+    }
+    break;  // no match
+  }
+
+  if (absl::ConsumePrefix(&s, "ago")) {
+    ConsumeSpaces(s);
+    intvl = -intvl;
+  }
+
+  if (!s.empty()) {
+    return internal::InvalidArgumentError(
+        absl::StrCat("\"", s_, "\": Syntax error at \"", s.substr(0, 5), "\""),
+        GCP_ERROR_INFO());
+  }
+  return intvl;
+}
+
+}  // namespace
+
+bool operator==(Interval const& a, Interval const& b) {
+  return Cmp<std::equal_to>(a.months_, a.days_, a.offset_,  //
+                            b.months_, b.days_, b.offset_);
+}
+
+bool operator<(Interval const& a, Interval const& b) {
+  return Cmp<std::less>(a.months_, a.days_, a.offset_,  //
+                        b.months_, b.days_, b.offset_);
+}
+
+Interval Interval::operator-() const {
+  Interval intvl;
+  intvl.months_ = -months_;
+  intvl.days_ = -days_;
+  intvl.offset_ = -offset_;
+  return intvl;
+}
+
+Interval& Interval::operator+=(Interval const& intvl) {
+  months_ += intvl.months_;
+  days_ += intvl.days_;
+  offset_ += intvl.offset_;
+  return *this;
+}
+
+Interval& Interval::operator*=(double d) {
+  // Fractional results only flow down into smaller units. Nothing ever
+  // carries up into larger units. This means that '1 month' / 2 becomes
+  // '15 days, but '1 month 15 days' * 3 is '3 months 45 days'.
```

Comment:
Only because that is how the operation is defined (to the best of my understanding).  The important takeaway is that the months/days/duration pieces should be kept distinct (seeing as they don't really inter-convert), until you have no choice.  For example, addition and negation are field-by-field, but the alternative here would be that `"1 month" / 2` is zero, and I guess that was deemed worse.

Reply:
If there is a go/ link or something, maybe add that as a comment. Otherwise, SGTM.

Reply:
That particular example, albeit written as `"1 month" * 0.5`, comes from https://gist.github.com/henryivesjones/ebd653acbf61cb408380a49659e2be97, which is cited near the test of that case.

Internally there is a `"1 day" / 2` -> `"12 hours"` example in go/spangres-interval-design, but I hesitate to mention that in the code.